File: matiasmasso/Mat.NET/DT2/DTOStaff.vb
Public Class DTOStaff
    Inherits DTOContact

    Property abr As String
    <JsonIgnore> Property avatar As Image

    Property alta As Date
    Property baixa As Date
    Property birth As Date
    Property sex As DTOEnums.Sexs
    Property numSs As String
    Property iban As DTOIban
    Property staffPos As DTOStaffPos

    Property category As DTOStaffCategory

    Public Sub New()
        MyBase.New()
    End Sub

    Public Sub New(oGuid As Guid)
        MyBase.New(oGuid)
    End Sub

    Shared Function FromContact(oContact As DTOContact) As DTOStaff
        Dim retval As DTOStaff = Nothing
        If oContact Is Nothing Then
            retval = New DTOStaff
        Else
            retval = New DTOStaff(oContact.Guid)
            With retval
                .Emp = oContact.Emp
                .Nom = oContact.Nom
                .NomComercial = oContact.NomComercial
                .Nif = oContact.Nif
                .Address = oContact.Address
                .ContactClass = oContact.ContactClass
                .Lang = oContact.Lang
                .Rol = oContact.Rol
            End With
        End If
        Return retval
    End Function


    Shared Function Posicio(oStaff As DTOStaff, oLang As DTOLang) As String
        Dim retval As String = ""
        If oStaff IsNot Nothing Then
            If oStaff.StaffPos IsNot Nothing Then
                retval = DTOStaffPos.Nom(oStaff.StaffPos, oLang)
            End If
        End If
        Return retval
    End Function

    Shared Function Categoria(oStaff As DTOStaff) As String
        Dim retval As String = ""
        If oStaff IsNot Nothing Then
            If oStaff.Category IsNot Nothing Then
                retval = oStaff.Category.Nom
            End If
        End If
        Return retval
    End Function

    Shared Function SegSocialGrup(oStaff As DTOStaff) As String
        Dim retval As String = ""
        If oStaff IsNot Nothing Then
            If oStaff.Category IsNot Nothing Then
                If oStaff.Category.SegSocialGrup IsNot Nothing Then
                    retval = oStaff.Category.SegSocialGrup.Nom
                End If
            End If
        End If
        Return retval
    End Function

    Shared Function AliasOrNom(oStaff As DTOStaff) As String
        Dim retval As String = oStaff.Abr
        If retval = "" Then retval = oStaff.Nom
        If retval = "" Then retval = oStaff.FullNom
        Return retval
    End Function

    Public Function Age(Optional toFch As Date = Nothing) As Integer
        Dim retval As Integer = 0

        If _birth <> Nothing Then
            If toFch = Nothing Then toFch = Date.Today
            retval = toFch.Year - _birth.Year
            'Go back to the year the person was born in case of a leap year
            If (_birth.Date > toFch.AddYears(-retval)) Then
                retval -= 1
            End If
        End If
        Return retval
    End Function

    Shared Function Excel(oStaffs As List(Of DTOStaff), oLang As DTOLang, Optional oExercici As DTOExercici = Nothing) As MatHelperStd.ExcelHelper.Sheet
        Dim sSheetName As String = "staff"
        If oExercici IsNot Nothing Then sSheetName = oExercici.year
        Dim retval As New MatHelperStd.ExcelHelper.Sheet(sSheetName, "M+O Staff")
        With retval
            .AddColumn("Nom", MatHelperStd.ExcelHelper.Sheet.NumberFormats.PlainText)
            .AddColumn("NIF", MatHelperStd.ExcelHelper.Sheet.NumberFormats.W50)
            .AddColumn("Num SS", MatHelperStd.ExcelHelper.Sheet.NumberFormats.W50)
            .AddColumn("Alta", MatHelperStd.ExcelHelper.Sheet.NumberFormats.DDMMYY)
            .AddColumn("Baixa", MatHelperStd.ExcelHelper.Sheet.NumberFormats.DDMMYY)
            If oExercici IsNot Nothing Then
                .AddColumn("Dies " & oExercici.year, MatHelperStd.ExcelHelper.Sheet.NumberFormats.Integer)
            End If
            .AddColumn("Home", MatHelperStd.ExcelHelper.Sheet.NumberFormats.Integer)
            .AddColumn("Dona", MatHelperStd.ExcelHelper.Sheet.NumberFormats.Integer)
            .AddColumn("Neixament", MatHelperStd.ExcelHelper.Sheet.NumberFormats.DDMMYY)
            .AddColumn("Edat", MatHelperStd.ExcelHelper.Sheet.NumberFormats.Integer)
            .AddColumn("Grup de Cotització", MatHelperStd.ExcelHelper.Sheet.NumberFormats.PlainText)
            .AddColumn("Categoría Professional", MatHelperStd.ExcelHelper.Sheet.NumberFormats.PlainText)
            .AddColumn("Posició", MatHelperStd.ExcelHelper.Sheet.NumberFormats.PlainText)
            .AddColumn("Adreça", MatHelperStd.ExcelHelper.Sheet.NumberFormats.PlainText)
            .AddColumn("Codi postal", MatHelperStd.ExcelHelper.Sheet.NumberFormats.PlainText)
            .AddColumn("Població", MatHelperStd.ExcelHelper.Sheet.NumberFormats.PlainText)
            .AddColumn("Telefon", MatHelperStd.ExcelHelper.Sheet.NumberFormats.PlainText)
            .AddColumn("Iban", MatHelperStd.ExcelHelper.Sheet.NumberFormats.PlainText)
        End With

        For Each oStaff In oStaffs
            Dim oRow As MatHelperStd.ExcelHelper.Row = retval.AddRow
            With oRow
                .AddCell(oStaff.nom)
                .AddCell(oStaff.nif)
                .AddCell(oStaff.numSs)
                .AddCell(oStaff.alta)
                .AddCell(oStaff.baixa)
                If oExercici IsNot Nothing Then
                    .AddFormula("IF(ISBLANK(RC[-1]),""" & oExercici.year & "-12-31"",MIN(""" & oExercici.year & "-12-31"",RC[-1]))-MAX(""" & oExercici.year & "-01-01"",RC[-2])")
                End If
                .AddCell(If(oStaff.sex = DTOEnums.Sexs.Male, 1, 0))
                .AddCell(If(oStaff.sex = DTOEnums.Sexs.Female, 1, 0))
                .AddCell(oStaff.birth)
                If oStaff.birth = Nothing Then
                    .AddCell()
                Else
                    If oExercici Is Nothing Then
                        .AddCell(oStaff.Age())
                    Else
                        .AddFormula(oStaff.Age(New Date(oExercici.year, 1, 1)))
                    End If
                End If
                .AddCell(DTOStaff.SegSocialGrup(oStaff))
                .AddCell(DTOStaff.Categoria(oStaff))
                .AddCell(DTOStaff.Posicio(oStaff, oLang))
                oRow.AddCell(oStaff.address.text)
                oRow.AddCell(oStaff.address.zip.zipCod)
                oRow.AddCell(DTOAddress.LocationFullNom(oStaff.address, oLang))
                oRow.AddCell(oStaff.telefon)
                oRow.AddCell(DTOIban.Formated(oStaff.iban.digits))
            End With
        Next
        Return retval
    End Function


End Class

File: matiasmasso/Mat.NET/DTO/DTOStaff.cs
using Newtonsoft.Json;
using System;
using System.Collections.Generic;

namespace DTO
{
    public class DTOStaff : DTOContact
    {
        public string Abr { get; set; }
        [JsonIgnore]
        public Byte[] Avatar { get; set; }

        public DateTime Alta { get; set; }
        public DateTime Baixa { get; set; }
        public DateTime Birth { get; set; }
        public DTOEnums.Sexs Sex { get; set; }
        public string NumSs { get; set; }
        public DTOIban Iban { get; set; }
        public DTOStaffPos StaffPos { get; set; }

        public DTOStaffCategory Category { get; set; }
        public bool Teletrabajo { get; set; }

        public DTOStaff() : base()
        {
        }

        public DTOStaff(Guid oGuid) : base(oGuid)
        {
        }

        public static DTOStaff FromContact(DTOContact oContact)
        {
            DTOStaff retval = null;
            if (oContact == null)
                retval = new DTOStaff();
            else
            {
                retval = new DTOStaff(oContact.Guid);
                {
                    var withBlock = retval;
                    withBlock.Emp = oContact.Emp;
                    withBlock.Nom = oContact.Nom;
                    withBlock.NomComercial = oContact.NomComercial;
                    withBlock.Nifs = oContact.Nifs;
                    withBlock.Address = oContact.Address;
                    withBlock.ContactClass = oContact.ContactClass;
                    withBlock.Lang = oContact.Lang;
                    withBlock.Rol = oContact.Rol;
                }
            }
            return retval;
        }


        public static string Posicio(DTOStaff oStaff, DTOLang oLang)
        {
            string retval = "";
            if (oStaff != null)
            {
                if (oStaff.StaffPos != null)
                    retval = DTOStaffPos.Nom(oStaff.StaffPos, oLang);
            }
            return retval;
        }

        public static string Categoria(DTOStaff oStaff)
        {
            string retval = "";
            if (oStaff != null)
            {
                if (oStaff.Category != null)
                    retval = oStaff.Category.Nom;
            }
            return retval;
        }

        public static string SegSocialGrup(DTOStaff oStaff)
        {
            string retval = "";
            if (oStaff != null)
            {
                if (oStaff.Category != null)
                {
                    if (oStaff.Category.SegSocialGrup != null)
                        retval = oStaff.Category.SegSocialGrup.nom;
                }
            }
            return retval;
        }

        public static string AliasOrNom(DTOStaff oStaff)
        {
            string retval = oStaff.Abr;
            if (retval == "")
                retval = oStaff.Nom;
            if (retval == "")
                retval = oStaff.FullNom;
            return retval;
        }

        public int Age(DateTime toFch = default(DateTime))
        {
            int retval = 0;

            if (this.Birth != default(DateTime))
            {
                if (toFch == default(DateTime))
                    toFch = DTO.GlobalVariables.Today();
                retval = toFch.Year - this.Birth.Year;
                // Go back to the year the person was born in case of a leap year
                if ((this.Birth.Date > toFch.AddYears(-retval)))
                    retval -= 1;
            }
            return retval;
        }


        public static MatHelper.Excel.Sheet excel(List<DTOStaff> oStaffs, DTOLang oLang, DTOExercici oExercici = null/* TODO Change to default(_) if this is not a reference type */)
        {
            string sSheetName = "staff";
            if (oExercici != null)
                sSheetName = oExercici.Year.ToString();
            MatHelper.Excel.Sheet retval = new MatHelper.Excel.Sheet(sSheetName, "M+O Staff");
            {
                var withBlock = retval;
                withBlock.AddColumn("Nom", MatHelper.Excel.Cell.NumberFormats.PlainText);
                withBlock.AddColumn("NIF", MatHelper.Excel.Cell.NumberFormats.W50);
                withBlock.AddColumn("Num SS", MatHelper.Excel.Cell.NumberFormats.W50);
                withBlock.AddColumn("Alta", MatHelper.Excel.Cell.NumberFormats.DDMMYY);
                withBlock.AddColumn("Baixa", MatHelper.Excel.Cell.NumberFormats.DDMMYY);
                if (oExercici != null)
                    withBlock.AddColumn("Dies " + oExercici.Year, MatHelper.Excel.Cell.NumberFormats.Integer);
                withBlock.AddColumn("Home", MatHelper.Excel.Cell.NumberFormats.Integer);
                withBlock.AddColumn("Dona", MatHelper.Excel.Cell.NumberFormats.Integer);
                withBlock.AddColumn("Neixament", MatHelper.Excel.Cell.NumberFormats.DDMMYY);
                withBlock.AddColumn("Edat", MatHelper.Excel.Cell.NumberFormats.Integer);
                withBlock.AddColumn("Grup de Cotització", MatHelper.Excel.Cell.NumberFormats.PlainText);
                withBlock.AddColumn("Categoría Professional", MatHelper.Excel.Cell.NumberFormats.PlainText);
                withBlock.AddColumn("Posició", MatHelper.Excel.Cell.NumberFormats.PlainText);
                withBlock.AddColumn("Adreça", MatHelper.Excel.Cell.NumberFormats.PlainText);
                withBlock.AddColumn("Codi postal", MatHelper.Excel.Cell.NumberFormats.PlainText);
                withBlock.AddColumn("Població", MatHelper.Excel.Cell.NumberFormats.PlainText);
                withBlock.AddColumn("Telefon", MatHelper.Excel.Cell.NumberFormats.PlainText);
                withBlock.AddColumn("Iban", MatHelper.Excel.Cell.NumberFormats.PlainText);
            }

            foreach (var oStaff in oStaffs)
            {
                MatHelper.Excel.Row oRow = retval.AddRow();
                {
                    var withBlock = oRow;
                    withBlock.AddCell(oStaff.Nom);
                    withBlock.AddCell(oStaff.PrimaryNifValue());
                    withBlock.AddCell(oStaff.NumSs);
                    withBlock.AddCell(oStaff.Alta);
                    withBlock.AddCell(oStaff.Baixa);
                    if (oExercici != null)
                        withBlock.AddFormula("IF(ISBLANK(RC[-1]),\"" + oExercici.Year + "-12-31\",MIN(\"" + oExercici.Year + "-12-31\",RC[-1]))-MAX(\"" + oExercici.Year + "-01-01\",RC[-2])");
                    withBlock.AddCell(oStaff.Sex == DTOEnums.Sexs.male ? 1 : 0);
                    withBlock.AddCell(oStaff.Sex == DTOEnums.Sexs.female ? 1 : 0);
                    withBlock.AddCell(oStaff.Birth);
                    if (oStaff.Birth == default(DateTime))
                        withBlock.AddCell();
                    else if (oExercici == null)
                        withBlock.AddCell(oStaff.Age());
                    else
                        withBlock.AddFormula(oStaff.Age(oExercici.FirstFch()).ToString());
                    withBlock.AddCell(DTOStaff.SegSocialGrup(oStaff));
                    withBlock.AddCell(DTOStaff.Categoria(oStaff));
                    withBlock.AddCell(DTOStaff.Posicio(oStaff, oLang));
                    oRow.AddCell(oStaff.Address.Text);
                    oRow.AddCell(oStaff.Address.Zip.ZipCod);
                    oRow.AddCell(DTOAddress.LocationFullNom(oStaff.Address, oLang));
                    oRow.AddCell(oStaff.Telefon);
                    oRow.AddCell(DTOIban.Formated(oStaff.Iban.Digits));
                }
            }
            return retval;
        }
    }
}
